sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala

```diff
@@ -404,6 +405,16 @@ abstract class Expression extends TreeNode[Expression] {
     } else {
       ""
     }
+
+  /**
+   * A developer API for marking if an expression is likely to be expensive.
+   * The current only consumer of this is the pushdown optimizer.
+   * By default an expression is expensive if any of it's children are expensive.
+   */
+  @DeveloperApi
```

Comment:
We never add such annotation to catalyst plans/expressions, as they are very internal and we guarantee no binary compatibility.